### PR TITLE
[UPSERT] Check for conflict constraint errors within a transaction

### LINKED ISSUE #6738: Cannot UPSERT record created in the same transaction.
### What happens?

When I create an entity and then try to upsert it in the same transaction, it results in `Constraint Error: PRIMARY KEY or UNIQUE constraint violated: duplicated key`. When I try to upsert a record created in a different transaction, it succeeds. Expected behavior: upsert should succeed in both scenarios. 

### To Reproduce

Upsert in the same transaction fails: 
```SQL
CREATE TABLE test (a INTEGER PRIMARY KEY, b INTEGER);
BEGIN;
INSERT OR REPLACE INTO test (a, b) VALUES (1, 2);
INSERT OR REPLACE INTO test (a, b) VALUES (1, 3);
```
```
Error: Constraint Error: PRIMARY KEY or UNIQUE constraint violated: duplicated key
```

But if the record is created in another transaction, upsert succeeds: 

```SQL
INSERT OR REPLACE INTO test (a, b) VALUES (1, 2);
BEGIN;
INSERT OR REPLACE INTO test (a, b) VALUES (1, 3);
COMMIT;
SELECT * FROM test;
```

```
┌───────┬───────┐
│   a   │   b   │
│ int32 │ int32 │
├───────┼───────┤
│     1 │     3 │
└───────┴───────┘
```

### OS:

Linux

### DuckDB Version:

v0.7.1 b00b93f0b1

### DuckDB Client:

CLI

### Full Name:

Stepan Anokhin

### Affiliation:

Atlan

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

### PR DESCRIPTION
**Pull request by Tishj**
This PR fixes #6738, #7396

This PR also includes #7050 as both of them needed to extract the UNIQUE constraint verification logic, and on closer inspection I think that PR doesn't have a very big impact.

Summary of why this issue exists, what doesn't happen and now does happen:
> The issue here is that we create a transaction-local ART for the primary key Index, currently in the "on conflict do update .. | do nothing" handling code we only check the global ART, which does not detect a conflict yet.

> Then, because the conflicting tuple(s) are not filtered out, the constraint check done for the insert into the storage causes the exception you're seeing.

Now in UPSERT we check both the global storage (like we did before) and the transaction-local storage + ART.

### CODE REVIEW

**Comment by Tishj** [comment]
Since INSERT now checks all indexes (also explicitly created ones), the expected behavior for failing transactions changes a little, as a result `test/sql/index/art/art_issue_7349.test` needed to be edited, since the insert now already fails, before COMMIT.

File: src/storage/data_table.cpp [hunk 1]
```diff
@@ -530,6 +530,72 @@ void DataTable::VerifyNewConstraint(ClientContext &context, DataTable &parent, c
 	local_storage.VerifyNewConstraint(parent, *constraint);
 }
 
+bool HasUniqueIndexes(TableIndexList &list) {
+	bool has_unique_index = false;
+	list.Scan([&](Index &index) {
+		if (index.IsUnique()) {
+			return has_unique_index = true;
+			return true;
```
Comment (Mytherin) [inline src/storage/data_table.cpp:538]: Completely unrelated to this PR but this should probably be an enum not a bool in the API - the code is hard to understand (what do `true` and `false` mean). I might clean that up later.

File: test/sql/upsert/upsert_transaction.test [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+BEGIN TRANSACTION;
+
+statement ok
+INSERT INTO tbl VALUES
+	(2, 1),
+	(3, 1),
+	(4, 1);
+
+statement ok
+INSERT INTO tbl VALUES
```
Comment (Mytherin) [inline test/sql/upsert/upsert_transaction.test:47]: Could we check the return value of the inserts with ON CONFLICT clauses to verify the correct number of rows are returned?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! LGTM - some comments:

File: test/sql/upsert/upsert_transaction.test (hunk 2 above)
Comment (Tishj) [reply test/sql/upsert/upsert_transaction.test:47]: I think I did that, unless I'm misinterpreting your question?
line 83-98
```
# Both of these inserts turn into updates
# The first affected tuple (5) is an update in the global storage
# The second affected tuple (6) is an update in the local storage
# The third affected tuple is an insert into the local storage
query I
INSERT INTO tbl VALUES (5,0), (6,0), (7,0) ON CONFLICT (a) DO UPDATE set b = excluded.b;
----
3

# The only affected tuple is the insert into the local storage
# (5,0) causes a constraint error in the global storage
# (3,0) causes a constraint error in the local storage
query I
INSERT INTO tbl VALUES (-1, 0), (5,0), (6,0) ON CONFLICT (a) DO NOTHING;
----
1
```

File: src/execution/operator/persistent/physical_insert.cpp [hunk 3]
```diff
@@ -256,19 +252,49 @@ idx_t PhysicalInsert::PerformOnConflictAction(ExecutionContext &context, DataChu
...
+}
+
 // TODO: should we use a hash table to keep track of this instead?
-void PhysicalInsert::RegisterUpdatedRows(InsertLocalState &lstate, const Vector &row_ids, idx_t count) const {
+void PhysicalInsert::RegisterUpdatedRows(InsertLocalState &lstate, const Vector &row_ids, idx_t count,
+                                         bool global) const {
 	// Insert all rows, if any of the rows has already been updated before, we throw an error
 	auto data = FlatVector::GetData<row_t>(row_ids);
+
+	// Are these row ids refering to the transaction-local storage or the global storage?
```
Comment (taniabogatsch) [inline src/execution/operator/persistent/physical_insert.cpp:294]: Small nit: this comment reads a bit odd; maybe rephrase it as a statement instead of a question?

File: src/execution/operator/persistent/physical_insert.cpp [hunk 4]
```diff
@@ -343,9 +363,85 @@ idx_t PhysicalInsert::OnConflictHandling(TableCatalogEntry &table, ExecutionCont
...
+	// Remove the conflicting tuples from the insert chunk
+	SelectionVector sel_vec(lstate.insert_chunk.size());
+	idx_t new_size =
+	    SelectionVector::Inverted(conflicts.Selection(), sel_vec, conflicts.Count(), lstate.insert_chunk.size());
+	lstate.insert_chunk.Slice(sel_vec, new_size);
+	lstate.insert_chunk.SetCardinality(new_size);
+	return updated_tuples;
+}
+
+idx_t PhysicalInsert::HandleLocalConflicts(TableCatalogEntry &table, ExecutionContext &context,
```
Comment (taniabogatsch) [inline src/execution/operator/persistent/physical_insert.cpp:379]: It seems like there is a lot of duplicate code in `HandleGlobalConflicts` and `HandleLocalConflicts`. Is it possible to unify that a bit more?

File: src/execution/operator/persistent/physical_insert.cpp [hunk 5]
```diff
@@ -343,9 +363,85 @@ idx_t PhysicalInsert::OnConflictHandling(TableCatalogEntry &table, ExecutionCont
...
+				if (!data[i]) {
+					// Only populate the selection vector with the tuples that did not meet the condition
+					sel.Append(i);
+				}
+			}
+			combined_chunk.Slice(sel.Selection(), sel.Count());
+			row_ids.Slice(sel.Selection(), sel.Count());
+			DataTable::VerifyUniqueIndexes(local_storage.GetIndexes(data_table), context.client, lstate.insert_chunk,
+			                               nullptr);
+			throw InternalException("The previous operation was expected to throw but didn't");
```
Comment (taniabogatsch) [inline src/execution/operator/persistent/physical_insert.cpp:438]: I am curious as to why we need to run this again. Do you know if this is only relevant for debugging purposes? Especially because it throws an internal exception, triggering this with user input should not be possible. If it is for debugging only, maybe we should add `#ifdef DEBUG`? Same for the code in `HandleGlobalConflicts`.

**Review by taniabogatsch — CHANGES_REQUESTED** [review]
It's nice to see that it was possible to 'just' add another constraint-checking step against the thread-local ART to fix this! I left a few minor comments. :)

File: src/execution/operator/persistent/physical_insert.cpp (hunk 5 above)
Comment (Tishj) [reply src/execution/operator/persistent/physical_insert.cpp:438]: We've found a conflict, we checked the condition that it should meet to be picked up by the DO UPDATE|NOTHING and it doesnt meet it.

So this should throw an error.
Because we would only get here if we found a conflict, we can safely rerun the check to get the error.

The internal exception should never happen because that code should be unreachable, I'd prefer to verify that that is the case on release as well

File: test/sql/upsert/upsert_transaction.test (hunk 2 above)
Comment (Mytherin) [reply test/sql/upsert/upsert_transaction.test:47]: Ah fair, I meant adding it to all inserts but perhaps that is overkill

**Comment by Mytherin** [comment]
@taniabogatsch could you have another look at this PR?

File: src/execution/operator/persistent/physical_insert.cpp [hunk 6]
```diff
@@ -256,19 +259,44 @@ idx_t PhysicalInsert::PerformOnConflictAction(ExecutionContext &context, DataChu
...
 	executor.Execute(chunk, update_chunk);
 	update_chunk.SetCardinality(chunk);
+}
+
+template <bool GLOBAL>
+static idx_t PerformOnConflictAction(ExecutionContext &context, DataChunk &chunk, TableCatalogEntry &table,
+                                     Vector &row_ids, const PhysicalInsert &op) {
+
+	auto &action_type = op.action_type;
+	if (action_type == OnConflictAction::NOTHING) {
```
Comment (taniabogatsch) [inline src/execution/operator/persistent/physical_insert.cpp:269]: Nit: we don't use the variable `action_type` anywhere else, no?

File: src/include/duckdb/execution/operator/persistent/physical_insert.hpp [hunk 7]
```diff
@@ -112,13 +112,21 @@ class PhysicalInsert : public PhysicalOperator {
 	                            ExpressionExecutor &defaults_executor, DataChunk &result);
 
 protected:
+	// idx_t HandleGlobalConflicts(TableCatalogEntry &table, ExecutionContext &context, InsertLocalState &lstate,
+	//                            DataTable &data_table) const;
+	// idx_t HandleLocalConflicts(TableCatalogEntry &table, ExecutionContext &context, InsertLocalState &lstate,
+	//                           DataTable &data_table) const;
```
Comment (taniabogatsch) [inline src/include/duckdb/execution/operator/persistent/physical_insert.hpp:118]: Commented out code.

File: src/include/duckdb/execution/operator/persistent/physical_insert.hpp [hunk 8]
```diff
@@ -112,13 +112,21 @@ class PhysicalInsert : public PhysicalOperator {
...
+	                       DataChunk &result) const;
 	idx_t OnConflictHandling(TableCatalogEntry &table, ExecutionContext &context, InsertLocalState &lstate) const;
-	idx_t PerformOnConflictAction(ExecutionContext &context, DataChunk &chunk, TableCatalogEntry &table,
-	                              Vector &row_ids) const;
-	void RegisterUpdatedRows(InsertLocalState &lstate, const Vector &row_ids, idx_t count) const;
+	// idx_t PerformGlobalOnConflictAction(ExecutionContext &context, DataChunk &chunk, TableCatalogEntry &table,
+	//                                    Vector &row_ids) const;
+	// idx_t PerformLocalOnConflictAction(ExecutionContext &context, DataChunk &chunk, TableCatalogEntry &table,
+	//                                   Vector &row_ids) const;
+	// void RegisterUpdatedRows(InsertLocalState &lstate, const Vector &row_ids, idx_t count, bool global) const;
```
Comment (taniabogatsch) [inline src/include/duckdb/execution/operator/persistent/physical_insert.hpp:129]: Commented out code.

**Review by taniabogatsch — CHANGES_REQUESTED** [review]
Looks almost good to me; I left some minor nits/comments. Very clean now, with `GLOBAL` and templating!

File: src/include/duckdb/execution/operator/persistent/physical_insert.hpp (hunk 7 above)
Comment (Tishj) [reply src/include/duckdb/execution/operator/persistent/physical_insert.hpp:118]: Ah oops, should have seen that myself

**Comment by Mytherin** [comment]
It's fine like this, thanks!
